### DISCUSSION TITLE
Parallel execution of scenarios

### DISCUSSION THREAD

**Discussion by mitchgrout** [cucumber/common#2184] [Ideas]
I'm using `cucumber` to provide BDD testing within C/C++ libraries that I'm working on; this typically involves creating a REPL executable around my library, and using `aruba` to send various queries to perform my tests.

The main issue I have is that my test executable, and `cucumber`, are both single-threaded, and so I often have a significant portion of CPU time left unspent. Similarly, some test executables tend to be IO-bound, leading to large waits between steps.

I am aware that `parallel_cucumber` can assist here by spawning a process **per feature**, but this has a few downsides:

1. Most reporters, including the `pretty` and `html` formatters, are unusable since each process attempts to write content to the same resource (`stdout`, or the report file) simultaneously
2. Since parallelism is performed **per feature**, there is no benefit when running a single feature in isolation

Looking at the source of `cucumber` and `cucumber-core`,  it seems feasible that `cucumber` itself could invoke scenarios in parallel, capturing the events and passing them to the underlying reporters in the correct fashion. C#'s SpecFlow also appears to have the same type of parallelism capabilities.

I'd be happy to work on implementing this for Ruby.

**Comment by mpkorstanje** [comment]
I reckon you would be best of creating a feature request at https://github.com/cucumber/cucumber-ruby.
